gobblin-utility/src/main/java/gobblin/util/request_allocation/ConcurrentBoundedPriorityIterable.java

```diff
@@ -0,0 +1,150 @@
+   * @return true if the element was added, false if there was no space and we could not evict any elements to make it fit.
+   *         Note that the element may get evicted by future offers, so a return of true is not a guarantee that the
+   *         element will be present at any time in the future.
+   */
+  public boolean add(T t) {
+    boolean addedWorkunits = addImpl(t);
+    if (!addedWorkunits) {
+      this.rejectedElement = true;
+    }
+    return addedWorkunits;
+  }
+
+  private synchronized boolean addImpl(T t) {
+    AllocatedRequestsBase.TAndRequirement<T> newElement = new AllocatedRequestsBase.TAndRequirement<>(t, this.estimator.estimateRequirement(t, this.resourcePool));
+
+    if (this.resourcePool.exceedsHardBound(newElement.getResourceRequirement(), false)) {
+      // item does not fit even in empty pool
+      return false;
+    }
+
+    ResourceRequirement candidateRequirement = ResourceRequirement.add(this.currentRequirement, newElement.getResourceRequirement(), null);
+
+    if (this.resourcePool.exceedsHardBound(candidateRequirement, false)) {
+      if (this.elements.size() == 0) {
+        // no elements, first element doesn't fit
+        return false;
+      }
+
+      if (this.comparator.compare(this.elements.last().getT(), t) <= 0) {
+        return false;
+      }
+
+      List<AllocatedRequestsBase.TAndRequirement<T>> toDrop = Lists.newArrayList();
+
+      ResourceRequirement tmpRequirement = this.currentRequirement.clone();
+      ResourceRequirement reuse = tmpRequirement.clone();
+
+      for (AllocatedRequestsBase.TAndRequirement<T> dropCandidate : this.elements.descendingSet()) {
+        if (this.comparator.compare(dropCandidate.getT(), t) <= 0) {
+          return false;
+        }
+        tmpRequirement.subtract(dropCandidate.getResourceRequirement());
+        toDrop.add(dropCandidate);
+        if (!this.resourcePool.exceedsHardBound(
+            ResourceRequirement.add(tmpRequirement, newElement.getResourceRequirement(), reuse), false)) {
+          break;
+        }
+      }
+
+      for (AllocatedRequestsBase.TAndRequirement<T> drop : toDrop) {
+        this.currentRequirement.subtract(drop.getResourceRequirement());
+        this.elements.remove(drop);
+      }
+    }
+
+    this.elements.add(newElement);
+    this.currentRequirement.add(newElement.getResourceRequirement());
+    return true;
+  }
+
+  /**
+   * @return Whether any calls to {@link #add} have returned false, i.e. some element has been rejected due
+   * to strict capacity issues.
+   */
+  public boolean hasRejectedElement() {
+    return this.rejectedElement;
+  }
+
+  /**
+   * @return Whether the list has reached its soft bound.
+   */
+  public synchronized boolean isFull() {
+    return this.resourcePool.exceedsSoftBound(this.currentRequirement, true);
+  }
+
+  @Override
+  public Iterator<AllocatedRequestsBase.TAndRequirement<T>> iterator() {
+    return this.elements.iterator();
```

Comment:
This is dangerous as the elements treeset can be modified concurrently. 

See also the javadoc for TreeSet iterators.

> Note that the fail-fast behavior of an iterator cannot be guaranteed as it is, generally speaking, impossible to make any hard guarantees in the presence of unsynchronized concurrent modification. Fail-fast iterators throw ConcurrentModificationException on a best-effort basis. Therefore, it would be wrong to write a program that depended on this exception for its correctness: the fail-fast behavior of iterators should be used only to detect bugs.